crates/iota-data-ingestion-core/src/progress_store/file.rs

```diff
@@ -95,11 +98,30 @@ impl FileProgressStore {
 
     /// Writes the given data to the file, overwriting any existing content.
     async fn write_to_file(&mut self, data: impl AsRef<[u8]>) -> IngestionResult<()> {
-        // before writing seek to the start of the file
-        self.file.seek(SeekFrom::Start(0)).await?;
-        // clear the file content
-        self.file.set_len(0).await?;
-        Ok(self.file.write_all(data.as_ref()).await?)
+        let tmp_path = self.path.with_extension("tmp");
+
+        {
+            let mut tmp_file = OpenOptions::new()
+                .write(true)
+                .create(true)
+                .truncate(true)
+                .open(&tmp_path)
+                .await?;
+            tmp_file.write_all(data.as_ref()).await?;
+
+            // only for testing add a small delay, useful for simulate crashes
+            if cfg!(test) {
+                tokio::time::sleep(std::time::Duration::from_nanos(10)).await;
+            }
+        }
+
+        // Atomically replace the original file
+        tokio::fs::rename(&tmp_path, &self.path).await?;
+
+        // Re-open the file handle for further reads
+        self.file = Self::open_or_create_file(&self.path).await?;
```

Comment:
At this point we know that the file exists, so why not open it directly and not use `open_or_create_file`?

It would avoid breaking the signature of `open_or_create_file` above.

Reply:
Mostly because I wanted to reuse the already present logic, but thinking now it can be misleading. Will follow with a patch shortly.

Reply:
fixed: https://github.com/iotaledger/iota/pull/6918/commits/3652ae03b3c808a3a41cb0ad11d58eab9a2938ff